File: third-semester/homework1.1/LazyEvaluation/MultiThreadedLazy.cs
namespace LazyEvaluation
{
    using System;

    /// <summary>
    /// Thread-safe lazy evaluation implementation
    /// </summary>
    /// <typeparam name="T">type of evaluation result</typeparam>
    public class MultiThreadedLazy<T> : ILazy<T>
    {
        private readonly object lockObject = new object();
        private volatile bool isEvaluated = false;
        private Func<T> supplier;
        private T result;

        /// <summary>
        /// Initializes a new instance of the <see cref="MultiThreadedLazy{T}"/> class.
        /// </summary>
        /// <param name="supplier">function that represents evaluation</param>
        public MultiThreadedLazy(Func<T> supplier)
            => this.supplier = supplier;

        /// <summary>
        /// Gets evaluation result
        /// </summary>
        /// <returns>evaluation result</returns>
        public T Get()
        {
            if (!this.isEvaluated)
            {
                lock (this.lockObject)
                {
                    if (!this.isEvaluated)
                    {
                        this.result = this.supplier();
                        this.isEvaluated = true;
                        this.supplier = null;
                    }
                }
            }

            return this.result;
        }
    }
}

File: 4th-semester/Homework8.1/Lazy/MultiThreadedLazy.fs
namespace Lazy
open System

/// Thread-safe lazy locking implementation
type MultiThreadedLazy<'a>(supplier: unit -> 'a) =
    /// Lock object
    let lockObject = Object
    /// Evaluation result
    let mutable result = None
    
    interface ILazy<'a> with
        /// Gets evaluation result
        /// Supplier function is called only once
        member this.Get() =
            if result.IsNone
                then lock lockObject (fun () -> if result.IsNone
                                                    then result <- Some (supplier ()))
            
            result.Value

